File: cpp/113/Path_Sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> ret;
        vector<int> v;
        pathSum(root, sum, ret, v);
        return ret;
    }
    //深度优先搜索，遍历过程中记录路过节点的值，到达叶子节点时，将当前路径进行记录
    void pathSum(TreeNode* root, int sum, vector<vector<int>> &ret, vector<int> &v){
        if(root){
            v.push_back(root->val);
            if(root->left==NULL&&root->right==NULL&&sum==root->val){
                ret.push_back(v);
            }
            pathSum(root->left, sum-root->val, ret, v);
            pathSum(root->right, sum-root->val, ret, v);
            v.pop_back();
        }
    }
};

File: python/113/Path_SumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        res = []
        stack = []
        def dfs(root, sum):
            if root:
                stack.append(root.val)
                if root.left is root.right and root.val == sum:
                    # 拷贝一份新的stack
                    res.append(list(stack))
                dfs(root.left, sum - root.val) 
                dfs(root.right, sum - root.val)
                stack.pop()
        dfs(root, sum)
        return res
